src/buf_writer.c

```diff
@@ -56,24 +61,25 @@ buf_writer_init(tsdn_t *tsdn, buf_writer_t *buf_writer, write_cb_t *write_cb,
 		    buf_len);
 		buf_writer->internal_buf = true;
 	}
-	buf_writer->buf_end = 0;
 	if (buf_writer->buf != NULL) {
 		buf_writer->public_write_cb = buf_writer_cb;
 		buf_writer->public_cbopaque = buf_writer;
 		buf_writer->private_write_cb = write_cb;
 		buf_writer->private_cbopaque = cbopaque;
 		buf_writer->buf_size = buf_len - 1; /* Allowing for '\0'. */
-		buf_writer_assert(buf_writer);
-		return false;
 	} else {
 		buf_writer->public_write_cb = write_cb;
 		buf_writer->public_cbopaque = cbopaque;
 		buf_writer->private_write_cb = NULL;
 		buf_writer->private_cbopaque = NULL;
 		buf_writer->buf_size = 0;
-		buf_writer_assert(buf_writer);
-		return true;
 	}
+	buf_writer->buf_end = 0;
+	buf_writer->allow_auto_flush = true;
```

Comment:
nit: let's do `_set_allow_auto_flush(..., false)` to make the default clear. We'd also need to move the `buf_writer_assert` in set_allow_auto_flush to its end.

Reply:
Defined a `buf_writer_set_allow_auto_flush_internal()` to avoid duplicating the logic without sacrificing the initial `buf_writer_assert()` in `buf_writer_set_allow_auto_flush()`, which I think is important for catching any bugs as early as possible and serving as a pre-condition verification.

Also I think having another `buf_writer_assert()` at the end of `buf_writer_set_allow_auto_flush()` makes sense. And I also added it to other similar places.